[RHELC-597] Register with subscription-manager via dbus

abadger opened the pull request:
Part of [RHELC-597](https://issues.redhat.com/browse/RHELC-597)

---

Using the dbus API to register, we're able to stop passing the
activation key over the commandline.

This also lets us avoid pexpect for invoking subscription-manager when
passing the password.

Note that this requires that dbus is installed and running and that the
dbus-python/python-dbus packages are installed.  This package is
available on RHEL6-RHEL9 clones (all that we care about) but is a new
dependency.  The subscription-manager rpm already requires dbus-python so when we install that, we should be installing sbud as well.  RHEL7 and above use systemd which requires that dbus be running.  So RHEL6 is the only place where we might have to install and start up DBus.  Since the supported migration would include upgrading RHEL6 to RHEL7 soon after conversion, we might consider it not worthwhile to stop and remove dbus in this case.  Still have to consider rollback.

This code has been successfully tested on centos8 with activation key.  There are quite a few questions that I'm ready to take some feedback on before going further:

## TODOs

- [ ] DBus Daemon
  - [x] For this release and PR we will 
    - [x] Add an rpm requires for the dbus daemon package
    - [x] Check whether dbus is running:
      - [x] systeminfo.py will record whether dbus is running when convert2rhel starts
      - [x] checks.py will have a new check that fails if the dbus daemon is not running. 
    - [x] If it's not running error and ask the user to start it
  - [ ] I will open a jira ticket to make this behaviour better for the next release:
    - [ ] If the daemon is not running, start it.
    - [ ] During rollback, if the daemon was started by us, stop it.
- [ ] Remove RegistrationCommand.args
  - [x] Leave this in for this release as it isa handy way to log what we are using to talk to subscription-manager 
  - [ ] Will open a jira ticket to remove the function in a later PR.
    - [ ] Will change the logging to enumerate what we call subscription-manager's DBus API with
- [x] Workaround python-dbus broken logging (it initializes the root logger with a Handler that outputs to stderr. This causes our messages to be displayed twice and worse, if the user gives us data which can provoke an error when marshalling data to send to subscription-manager over DBus, then DBus will log the credentials in clear to stderr.)
  - [x] Added a NullHandler to the root logger before importing DBus which should stop DBus from adding the stderr handler.
  - [ ] Will open a jira ticket to move the NullHandler addition to another location next release.  @r0x0d is working on an early-initialization step for the yum merge transaction PR which we can use for this as well.
- [ ] Open jira ticket to determine org even for username + password auth
  - Resulted from speaking with Pino.
  - One customer bz: https://bugzilla.redhat.com/show_bug.cgi?id=1946309
  - Michal said: "We still can't handle this use case. For that reason we state in the product documentation that Satellite users should use the activation key/organization combination."
  - This code should be easy to update to get the organization: https://github.com/oamg/convert2rhel/blob/main/convert2rhel/subscription.py#L208
- [ ] Open a jira ticket to enter rollback if a user hits Ctrl-C during registration.  As coded right now, it will retry three times before entering rollback.
- [ ] Open a jira ticket to look into why using DBus to edit the configuration is not working at the end of `RegistrationCommand.__call__()`.  Looks to be dbus policy-related.
- [ ] Open a jira ticket to figure out:
  - Bug in subscription-manager that prevents force from working when we register when the system is already registered.  WIll this be fixed and can we drop the code that works around it?
  - Why we need to stop the rhsm service. Can we limit the rhsm service stop to only some platforms (RHEL7 and older)? 
  - Could we use reload instead of stop?  Do we need to stop even if unregister succeeds?
  - Some experimentation shows that setting register_opts force: True is having an effect.  Maybe the bug was that we weren't setting the DbusDictionary signature to `sv` and the sub-man server really is doing the right thing?
- [ ] Open a jira ticket to make sub-man config rollback aware. 
- [x] What should the return value of RegistrationCommand be?
  - [x] With Pino and Michal's feedback I lean towards RegistrationCommand returning nothing.
  - [x] On error, raise an exception.
  - [x] Be sure to catch the exception in our registration retry code.
- [x] If we fallback to cli, do we still allow activation_key even though it is insecure?
  - Decided not to fallback to cli since that would mean two code paths to test. 
- [x] Similar questions for python-dbus and dbus-python being installed.
  * The bindings are available on pypi as `dbus-python`.
  - [x] This is now required by the rpm.
- [x] Do we want to use DBUS for registering via passwords or only for activation key?
  - Decided to use DBus for registering via both passwords and activation key.
- [x] In _call_via_dbus(), we need to add more logging.  Call_via_cli lets run_subprocess()/run_in_pty() do most of its logging but we don't have that in _call_via_dbus()
- [x] integration tests to check that activation_key is not visible to the process list (because we're no longer calling subscription-manager with activation_key on the cli.
- [x] A few questions about schema, username, and password for pino.
  - [x] Open a jira ticket to look into what subscription-manager cli does with username:password in the --server-url and change our code to do what it does (our code currently ignores )
  - [x] Pino let us know that any username and password supplied to the subscription-manager cli via --serverurl is ignored.  So we can safely ignore those too.
- [x] unittests
  - [x] Fix current tests that fail after moving to dbus
  - [x] Write new tests to cover new code paths. 
- [x] We're going to keep dbus-python as a Requires in the rpm package so we don't need to conditionalize import
  - [ ]  Do we need dbus-python it in some requirements files?  we don't have pexpect in any requirements.

ptoscano commented on convert2rhel/subscription.py:
```diff
@@ -278,6 +287,93 @@ def args(self):
+        # Create a new bus so we can talk to rhsm privately (For security:
+        # talking on the system bus might be eavesdropped in certain scenarios)
+        register_server = system_bus.get_object("com.redhat.RHSM1", "/com/redhat/RHSM1/RegisterServer")
+        address = register_server.Start("", dbus_interface="com.redhat.RHSM1.RegisterServer")
```
the first parameter is the locale for the potential error messages; maybe it'd be better to pass the current locale of convert2rhel; the same applies to the other calls that take a locale as parameter

abadger replied:
Cool.  We always use the C locale for the things that we're calling so I will definitely add that to these function calls.

abadger replied:
Pushed a commit with all locale fields filled in.

ptoscano commented on convert2rhel/subscription.py:
```diff
@@ -278,6 +287,93 @@ def args(self):
+                    self.organization or "",
+                    self.username,
+                    self.password,
+                    "",
```
`options` is a dict, so this must be an empty dict rather than an empty string

abadger replied:
Pushed a change that changes this to `dbus.Dictionary({}, signature="ss")`

ptoscano commented on convert2rhel/subscription.py:
```diff
@@ -278,6 +287,93 @@ def args(self):
+                    self.password,
+                    "",
+                    connection_opts,
+                    "C",
```
ditto (see earlier notes wrt locale)

abadger replied:
Now uses a locale of `C`.

ptoscano commented on convert2rhel/subscription.py:
```diff
@@ -278,6 +287,93 @@ def args(self):
+        if self.server_url:
+            connection_opts = {"host": self.server_url}
+        else:
+            # DBus seems to want empty string if there are no values even though the type is a dict
+            connection_opts = ""
```
`connection_options` is a dict, so this must be an empty dict (it is bizzarre that it works, however it shouldn't)

abadger replied:
Pushed a change that changes this to `dbus.Dictionary({}, signature="ss")`.

Pino and I talked about this on chat.  An empty python dict doesn't work because dbus can't get enough introspection information from that to properly type the keys and values: https://gitlab.freedesktop.org/dbus/dbus-python/-/issues/25 .  `{"": ""}` works for `/bin/dbus-send` but it doesn't work in python-dbus for an unknown reason.

Manually creating the empty DBus Dictionary with signature information should work and better documents what the field means than empty string.

ptoscano commented on convert2rhel/subscription.py:
```diff
@@ -278,6 +287,93 @@ def args(self):
+                register_object.RegisterWithActivationKeys(
+                    self.org,
+                    [self.activation_key],
+                    "",
```
`options` is a dict, so this must be an empty dict rather than an empty string

abadger replied:
Pushed a change that changes this to dbus.Dictionary({}, signature="ss")

ptoscano commented on convert2rhel/subscription.py:
```diff
@@ -278,6 +287,93 @@ def args(self):
+                    [self.activation_key],
+                    "",
+                    connection_opts,
+                    "C",
```
ditto (see earlier notes wrt locale)

ptoscano commented on convert2rhel/subscription.py:
```diff
@@ -278,6 +287,93 @@ def args(self):
+
+        finally:
+            # Always shut down the private bus
+            register_server.Stop("", dbus_interface="com.redhat.RHSM1.RegisterServer")
```
ditto (see earlier notes wrt locale)

abadger replied:
Now uses the `C` locale.

ptoscano reviewed (commented):
I left some easy notes on the code; I can try provide some answers to your question, at least from the subscription-manager POV:

> Note that this requires that dbus is installed and running and that the dbus-python/python-dbus packages are installed. This package is available on RHEL6-RHEL9 clones (all that we care about) but is a new dependency.

New for convert2rhel, not for subscription-manager.

>     * [ ]  Should we check whether dbus is running?

As you wrote in the comments: RHEL 7 uses systemd, which mandates D-Bus, while RHEL 6 uses upstart.

Maybe it would be better to try to connect to the system bus, and in case start the D-Bus system service (it should affect only RHEL 6).

>     * [ ]  If it's not running should we install and run it or should we fall back to subscription-managr cli?
>     * [ ]  If we fallback to cli, do we still allow activation_key even though it is insecure?
>     * [ ]  What should the return value of RegistrationCommand be?  Should cli non-zero retcodes raise an exception instead of returning the code?
>     * [ ]  Do we want to use DBUS for registering via passwords or only for activation key?

Personally, I would switch the implementation of registration with subscription-manager to D-Bus exclusively. Otherwise you need to maintain two different code paths (D-Bus and cli), and given that D-Bus should be used when converting RHEL 7 and greater versions, it means the cli version will be tested less.

Also, using D-Bus exclusively means that potentially more interactions with subscription-manager could be done via D-Bus in the future, rather than spawning shell commands (which is a not so good idea, as the console output of subscription-manager is neither stable nor machine parseable).

>     * [ ]  Should checking for DBUS be done in either system_info or checks.py rather than in subscription.py?
> 
Since the D-Bus way also works with subscription-manager in RHEL 6, I'd rather error out in the subscription module if D-Bus is not available (cannot be installer or started). This way it will only matter if the subscription module is used.

Of course this depends on how much effort you want to invest in supporting RHEL 6->7 upgrades.

>       * We could inform the user of the problem and let them install and start dbus like we make them install a compatible kernel and reboot.

Can be a way.

>     * [ ]  If we install it and start the daemon, do we need to revert that after the upgrade/rollback?

I'd say no, as you will get anyway a D-Bus daemon installed in RHEL >= 7.

>     * [ ]  Similar questions for python-dbus and dbus-python being installed.
>     * [ ]  We could have the rpm require dbus and python-dbus which will make sure it is installed.  It won't make sure that dbus is running, though.

I'd recommend to add `python-dbus`/`python3-dbus` explicitly as dependency in the packaging (`convert2rhel.spec`): even if the subscription module is disabled, the Python module is very small in size and dependencies, so it should not be a problem to unconditionally require it even if the subscription module is disabled.

abadger commented:
>>    * [ ]  Should checking for DBUS be done in either system_info or checks.py rather than in subscription.py?
> Since the D-Bus way also works with subscription-manager in RHEL 6, I'd rather error out in the subscription module if D-Bus is not available (cannot be installer or started). This way it will only matter if the subscription module is used.

IIRC, subscription always happens.

If we are going to fail if dbus is not running, we would definitely want that to happen in the checks phase so as little as possible has happened to the system (which would mean we don't have any changes we have to rollback).  If we're going to try to install dbus for the user, then we still might want to find that fact out in system_info and then make use of that to decide what to do in subscription.py.

ptoscano reviewed (commented):
In addition to the inline comments I left, `RegistrationCommand.args()` is no more needed now, and can be dropped.

abadger commented on convert2rhel/subscription.py:
```diff
@@ -279,26 +288,89 @@ def args(self):
+        system_bus = dbus.SystemBus()
+
+        if self.server_url:
+            connection_opts = {"host": self.server_url}
```
Hey @ptoscano I'm back from some PTO.  Should protocol be discarded and user:password be given as parameters in the `Register()` dbus message?  (So only host, port, and path portions of the url will be passed as connection_opts?)

ptoscano replied:
> Should protocol be discarded

Yes.

> user:password be given as parameters in the Register() dbus message?

Yes.

> (So only host, port, and path portions of the url will be passed as connection_opts?)

Yes.

abadger replied:
Another related question, if the subscription-manager cli is given a username/password pair in both cli arguments and as part of the server-url, which takes precedence?  If they are only in the server-url, does the cli make use of it? If they are in the url but activation-key is given on the cli, are the username and password used?  (I'll probably try to model current behaviour so knowing these things will help me get that right).

ptoscano replied:
Hmm I need to check myself too, however for now I'd ignore the userinfo part of the server URL.

abadger commented on convert2rhel/subscription.py:
```diff
@@ -279,26 +288,89 @@ def args(self):
+        loggerinst.debug("Getting a subscription-manager RegisterServer object from dbus")
+        register_server = system_bus.get_object("com.redhat.RHSM1", "/com/redhat/RHSM1/RegisterServer")
+        loggerinst.debug("Starting a private DBus to talk to subscription-manager")
+        address = register_server.Start("C", dbus_interface="com.redhat.RHSM1.RegisterServer")
```
I like the idea of moving the locale out to a separate file in case we do localize convert2rhel in the future. I'm a little hesitant about changing this to use a non-C locale for several reasons, though:

* convert2rhel isn't localized itself.  Emitting a few localized messages while everything else is unlocalized doesn't seem like a good experience.
* In the past I've had some people make the point that error messages, are better left unlocalized so that people can easily put the errors into google (or bug reports).  I'm not quite convinced either way about that but I think it's appropriate to punt on that question until we decide to localize convert2rhel itself.

ptoscano replied:
> I like the idea of moving the locale out to a separate file

Oh sorry, I did not mean a new file on its own, rather a simple variable declared at the top of the `__call__()` function, eg.
```python
    SUBMAN_LOCALE = "C"
```
and use it in all the calls; this way, whatever you want to do in the future, it's a one place change.

abadger replied:
I moved the locale constants out to their own file in this commit: f90e34f715f955fd4514f6bc07be070c6c16e786

abadger commented on convert2rhel/subscription.py:
```diff
@@ -279,26 +288,89 @@ def args(self):
         else:
-            # Warning: Currently activation_key can only be specified on the CLI. This is insecure
-            # but there's nothing we can do about it for now. Once subscription-manager issue:
-            # https://issues.redhat.com/browse/ENT-4724 is implemented, we can change both password
-            # and activation_key to use a file-based approach to passing the secrets.
-            output, ret_code = utils.run_subprocess([self.cmd] + self.args, print_cmd=False)
+            connection_opts = EMPTY_DBUS_DICT
```
I like this. For now, though, I've taken a tiny bit of this and combined it with the construction of connection_opts from server_url.  I don't think we'll need to do anything dynamic with connection_opts in the convert2rhel use case so it feels like it's better to move the boilerplate into a separate function.  But if we do need to set connection_opts to different values, I'll keep this in mind as a good strategy.

abadger commented on convert2rhel/subscription.py:
```diff
@@ -279,26 +288,89 @@ def args(self):
+        # * Should checking for DBUS be done in either system_info or checks.py?
+        # * What should the return value of this code be?
+
+        # DBus throws errors for pure empty dict and dbus-python doesn't like
```
Reworded to account for this.

abadger commented on convert2rhel/subscription.py:
```diff
@@ -22,6 +22,15 @@
 from collections import namedtuple
 from time import sleep
 
+
+try:
+    import dbus
+    import dbus.connection
+
+    DBUS = True
+except ImportError:
+    DBUS = False
+
```
<nod>  Added his and removing RegistrationCommaqnd.args to the list of things to do before this code is ready.

abadger replied:
Good point about args.  I'll think for a bit about whether to remove args and restructure the logging or keep args as a handy way to visualize that.

abadger replied:
Conditional import of dbus has now been removed

abadger commented on packaging/convert2rhel.spec:
```diff
@@ -41,12 +44,14 @@ Requires:       dnf
 Requires:       dnf-utils
 Requires:       grubby
 Requires:       python3-pexpect
+Requires:       python4-dbus
```
Thanks.  Fixed.

ptoscano commented on convert2rhel/subscription.py:
```diff
@@ -286,6 +288,22 @@ def args(self):
+        if self.server_url:
+            url_parts = urllib.parse.urlsplit(self.server_url, allow_fragments=False)
+            connection_opts = {"host": url_parts.hostname}
+            connection_opts = {"port": url_parts.port}
```
if no port was specified, there is no `port` in `url_part`

abadger replied:
Fixed

ptoscano commented on convert2rhel/subscription.py:
```diff
@@ -286,6 +288,22 @@ def args(self):
+
+            path = urllib.parse.urlunsplit(
+                (None, None, url_parts.path, url_parts.query, url_parts.fragment))
+            connection_opts = {"handler": path}
```
hm why not simply use `url_parts.path`? it is already `/path` as needed for "handler"

abadger replied:
So.... I was thinking that if query and fragment were specified in a url, the handler would be the place for those to be passed to the subscription-manager backend.
* Are they just not used in subscription-manager right now?  If so I think this is future-proofing so I think I'd keep it.
* Are they handled by some other connection_opt?  Let me know if so and I'll move them there.

ptoscano replied:
AFAICT, query and fragment are ignored right now. They are ignored after parsing URLs from the command lines, and there are no config knobs for them (sub-man does REST calls, so they would need to be passed to each).

Because of that, I'd simply ignore query and fragment altogether, as (if they would be ever needed) they would require additional work in sub-man.

abadger replied:
This has been updated to follow ptoscano's suggestion.  Rationale here: https://github.com/oamg/convert2rhel/pull/540#discussion_r943661175

bocekm commented on convert2rhel/subscription.py:
```diff
@@ -277,28 +288,104 @@ def args(self):
-            # https://issues.redhat.com/browse/ENT-4724 is implemented, we can change both password
-            # and activation_key to use a file-based approach to passing the secrets.
-            output, ret_code = utils.run_subprocess([self.cmd] + self.args, print_cmd=False)
+        # Questions:
+        # * Should we check whether dbus is running?
+        # * If it's not running should we install and run it or should we fall back to
+        #   subscription-managr cli?
+        # * If we install it, do we need to revert that after the upgrade/rollback?
+        # * Should checking for DBUS be done in either system_info or checks.py?
+        # * What should the return value of this code be?
```
@abadger, are there still some of the questions unanswered after talking to @ptoscano? I haven't looked at the discussion of you two much so here's how I'd see the answers:

> Should we check whether dbus is running?

Yes.

> If it's not running should we install and run it or should we fall back to subscription-managr cli?

We're already decided to add to spec, so it's installed. If not running, we should start it (no fall back to the sub-man cli).

> If we install it, do we need to revert that after the upgrade/rollback?

If the service was not running and we started it, we should stop the service upon rollback. Let's do that though in a separate PR, after we merge https://github.com/oamg/convert2rhel/pull/516.

> Should checking for DBUS be done in either system_info or checks.py?

I'd say I prefer the checks.py but it's not that important.

> What should the return value of this code be?

I believe you can drop the code in `register_system()` that check the output string (introduced in
https://github.com/oamg/convert2rhel/commit/311b0bd363e3a4b170c9ea981194603e8b4e3858#diff-51e9ff11d39778d3b26778b293fc350430a82e2feed0dab2938da7c0069ffdc6R117). Thus, you don't need to return any string from the `__call__()`.

Regarding the return code - when the registration fails the `Register`/`RegisterWithActivationKeys` services raise an exception, right? We've got the loop for 3 registration attempts and to be able to get to another attempt it seems we should catch the exception. Oh, I now see Pino mentions the same in a comment https://github.com/oamg/convert2rhel/pull/540/files#r923341266.

abadger replied:
<nod>  I think I've captured all of your (and Pino's) feedback in the checklist in this PR's description.  Let me know if I've missed anything.

I'll remove this comment in the code as it is getting outdated.

r0x0d commented on convert2rhel/subscription.py:
```diff
@@ -277,28 +287,103 @@ def args(self):
+            * RHEL7:
+                * dbus looks to be installed and run by default
+                * dbus-python-1.1.9 is available
+            * RHEL8:
+            * RHEL9:
```
State for RHEL 8 and RHEL 9 is the same? If so, I would group them together like: 

```
RHEL8 & RHEL9:
```

abadger replied:
Good point.  I didn't have a RHEL8 vm handy when I was filling this comment out.  I checked things just now and RHEL8 and RHEL9 are almost identical.  RHEL8 just has a slightly older version of python3-dbus so I've merged them as you suggested.

ptoscano commented on convert2rhel/checks.py:
```diff
@@ -698,3 +699,21 @@ def is_loaded_kernel_latest():
+
+    if tool_opts.no_rhsm:
+        logger.info("Skipping the check because we have been asked not to subscribe this system to rhsm.")
+        return None
```
minor: only `return` is enough

abadger replied:
Changed to use bare return (along with small refactor of the code) in this commit: a049b58f83122b4d83348ca25484db7c4d357d18

ptoscano commented on convert2rhel/checks.py:
```diff
@@ -698,3 +699,21 @@ def is_loaded_kernel_latest():
+        )
+
+    logger.info("DBus Daemon is running")
+    return None
```
minor: only `return` is enough, or even no `return` is needed here

abadger replied:
I was a bit on the fence about this one.  I kept the `None`s in because I am returning in two different levels but the counter argument is that we aren't using the return value from any of the `check_*` functions.  I've restructured this code so that the returns are both "exit early" returns at the top.  Using bare `return`s with that structure feel much better to me.

ptoscano commented on convert2rhel/subscription.py:
```diff
@@ -277,28 +287,103 @@ def args(self):
+            if url_parts.port is not None:
+                connection_opts["port"] = url_parts.port
+
+            # Current subscription-manager server-side does not take query and fragments at this
+            # path but include them for future proofing.
+            path = urllib.parse.urlunsplit(("", "", url_parts.path, url_parts.query, url_parts.fragment))
+            connection_opts["handler"] = path
```
I'd simply leave them out; it would require work in sub-man to support them, if ever, so let's keep things simple here and use only `url_parts.path`

abadger replied:
<nod> I can see your point of view as a subscription-manager dev.  From my point of view, though, I'd rather code something now that might work if subscription-manager should choose to start using query and fragment in the future while the way that subscription-manager and dbus work is fresh in my mind.  If subscription-manager never needs to add query or fragment here, then there is no harm.  If subscription-manager adds them in the way that the code anticipates, then I won't have to try to figure out the mapping of cli->dbus args again.  If subscription-manager adds them in a way that the code doesn't anticipate, at least when someone looks at the code they'll see all the pieces even if they're assembled in the wrong way.

ptoscano replied:
I don't think subscription-manager is going to get that functionality any time soon. It is not only about taking query and fragment from an URL passed to `--serverurl`, but also needs new configuration bits (as hostname, port, etc are stored separately there) and all the associated machinery for it, and there is simply nothing that ATM needs them.

So, while I understand your POV, I'd keep things simple and not adding things that could misbehave.

abadger commented on convert2rhel/unit_tests/subscription_test.py:
```diff
@@ -35,6 +38,41 @@
+
+@pytest.fixture
+def mocked_rhsm_register_with_activation_keys(monkeypatch, request, _mocked_rhsm_register_base):
+    if pytest.__version__.split(".") <= ["3", "6", "0"]:
+        rhsm_returns = request.node.get_marker("rhsm_returns")
+    else:
+        rhsm_returns = request.node.get_closest_marker("rhsm_returns")
```
I hear you, yeah.  What do you think about making a helper function that we can call with the name of the mark and the helper returns the mark to the fixture?

abadger replied:
I've created the helper here: 1d003704b4bb877b800341531be0d0a462fc7a29  I've ported the two test functions that use it in this PR to use it but I did not port the one that you are working on in the RHEL6 PR.

r0x0d replied:
Nice! I will port that out to my RHEL6 PR.

ptoscano commented on convert2rhel/subscription.py:
```diff
@@ -277,28 +287,103 @@ def args(self):
+        """
+        Use dbus to register the system with subscription-manager.
+
+        Status of dbus on various platforms:
+            * RHEL6:
+                * DBUS is present but may not be installed.
+                * Install the dbus package and run /etc/rc.d/init.d/messagebus start
+                * dbus-python-0.83.0 is available
+            * RHEL7:
+                * dbus looks to be installed and run by default
+                * dbus-python-1.1.9 is available
+            * RHEL8:
+            * RHEL9:
+                * dbus looks to be installed and run by default
+                * python3-dbus-1.2.18 is available
```
TBH I'm not sure how the version of dbus-python is useful in this context; I'd simplify thinks a bit as:
- RHEL 6:
  - D-Bus may not be installed or running by default
  - Install the 'dbus' package and run `/etc/rc.d/init.d/messagebus start`
- RHEL 7 and greater:
  - D-Bus is an hard system dependency because of systemd
  - a D-Bus daemon is installed and running

abadger replied:
I decided to leave the versions in because there were some features/bugfixes that I could not make use of since they were added in a later version than some versions of RHEL had.  Having a quick table of relevant versions on all the RHEL versions we're targetting was helpful for quickly verifying that we weren't going to run into a problem.  However, now that I think about it, those might have been versions of the dbus daemon rather than the dbus python bindings.  I'll add the dbus version to the information in this content as well.

ptoscano reviewed (commented):
Thanks for the various updates! At least from my (subscription-manager) POV, the implementation looks sane and matches the sub-man API.

I left some notes for small/minor things; nothing biggie though.

One open question I see in the big checklist:

> A few questions about schema, username, and password for pino. 

username & password specified in `--serverurl URL` are ignored; hence, you can safely ignore them too.

abadger commented on convert2rhel/subscription.py:
```diff
@@ -279,26 +288,89 @@ def args(self):
+            loggerinst.debug("Shutting down private DBus instance")
+            register_server.Stop("C", dbus_interface="com.redhat.RHSM1.RegisterServer")
+
+        return "", 0
```
This has now been changed to return nothing in the success case and raise an exception in the error case.

r0x0d commented on convert2rhel/systeminfo.py:
```diff
@@ -359,6 +360,35 @@ def corresponds_to_rhel_eus_release(self):
+            for line in output.splitlines():
+                if line.startswith("ActiveState="):
+                    state = line.split("=", 1)[1]
+                    if state in ("active", "reloading", "activating"):
```
Is there any possibility for the `reloading` or `activating` status to turn to be a failure in the end? 

I mean, I don't know if that's even possible, but, let's pretend that after reload something goes wrong, would that affect us?

abadger replied:
There is.  I thought about implementing retry on these but got lazy.  I've now pushed a commit which will retry three times if the state is not definitive (for systemd: active, failed, inactive, or deactivating)

ptoscano commented on convert2rhel/subscription.py:
```diff
@@ -385,6 +394,26 @@ def __call__(self):
+    # https://github.com/candlepin/subscription-manager/blob/main/src/rhsm/utils.py#L112
+
+    # It should look like it has a scheme followed by other information.
+    if not re.match("^.+://.+", server_url):
```
i'd use `[^:]+` instead of `.+`, to avoid potential greedyness of regexps

abadger replied:
Good idea.

abadger replied:
change pushed in a new commit.

ptoscano reviewed (commented):
also there's a typo `compatiility` in the commit message

abadger commented:
Okay, I think I've addressed all the comments that have been raised here and in our internal chat.  I have run my checkout against one OEL8 vagrant instance and I don't see any duplicated log messages in the run.

I don't like where it is that I am adding NullHandler to the root logger but I think it will do for now.  @r0x0d  is working on adding a dedicated setup step as part of his work on merging the yum transactions.  When that happens, I think moving the root log handler setup to there would make sense.

In the initial port for this PR, I have a list of checklist items.  All the ones left undone are things that I think are things I think we should change with this implementation but we can defer them to the next release.I'll open jira tickets for all of those.

ptoscano commented:
> I don't like where it is that I am adding NullHandler to the root logger but I think it will do for now.

From what I saw in the dbus-python code, the logged is initialized only when printing "severe problems" that happen when using dbus-python. For example, passing `None` as value for e.g. a string. subscription-manager uses dbus-python way more than this code, and TTBOMK we have not seen any problem related to this, neither as printed directly to terminal nor in the log file.
While I understand your POV on defensive programming, from what I see all the parameters are already checked in one way or another. So IMHO this extra commit could not be needed, now that the code is fixed.

SpyTec commented on convert2rhel/subscription.py:
```diff
@@ -277,28 +287,131 @@ def args(self):
+        if self.server_url:
+            try:
+                url_parts = parse_url(self.server_url)
+            except ValueError:
+                # url_parts will remain None which will mean we return an empty dbus.DIctionary
+                pass
```
Untested

abadger replied:
This branch is now tested.  But we're looking into some changes to this behaviour (error instead of pass).

SpyTec commented on convert2rhel/checks.py:
```diff
@@ -698,3 +699,21 @@ def is_loaded_kernel_latest():
+    logger.task("Prepare: Check that DBus Daemon is running")
+
+    if tool_opts.no_rhsm:
+        logger.info("Skipping the check because we have been asked not to subscribe this system to rhsm.")
```
```suggestion
        logger.info("Skipping the check because we have been asked not to subscribe this system to RHSM.")
```

SpyTec commented on convert2rhel/unit_tests/checks_test.py:
```diff
@@ -1215,3 +1218,43 @@ def test_is_loaded_kernel_latest_unsupported_skip(
+@pytest.mark.parametrize(
+    ("no_rhsm", "dbus_running", "log_msg"),
+    (
+        (True, True, "Skipping the check because we have been asked not to subscribe this system to rhsm."),
+        (True, False, "Skipping the check because we have been asked not to subscribe this system to rhsm."),
```
```suggestion
        (True, True, "Skipping the check because we have been asked not to subscribe this system to RHSM."),
        (True, False, "Skipping the check because we have been asked not to subscribe this system to RHSM."),
```

SpyTec commented on convert2rhel/unit_tests/checks_test.py:
```diff
@@ -1215,3 +1218,43 @@ def test_is_loaded_kernel_latest_unsupported_skip(
+    log_msg = (
+        "Could not find a running DBus Daemon which is needed to"
+        " register with subscription manager.\nPlease start dbus using `systemctl"
+        " start dbus` or (on CentOS 6), `service messagebus start`"
```
```suggestion
        " start dbus` or (on CentOS Linux 6), `service messagebus start`"
```

SpyTec commented on convert2rhel/checks.py:
```diff
@@ -698,3 +699,21 @@ def is_loaded_kernel_latest():
+
+    logger.critical(
+        "Could not find a running DBus Daemon which is needed to register with subscription manager.\n"
+        "Please start dbus using `systemctl start dbus` or (on CentOS 6), `service messagebus start`"
```
```suggestion
        "Please start dbus using `systemctl start dbus` or (on CentOS Linux 6), `service messagebus start`"
```

SpyTec commented on convert2rhel/subscription.py:
```diff
@@ -277,28 +287,131 @@ def args(self):
+
+
+def parse_url(server_url):
+    """Parse a string as a url in a manner mostly compatible with subscription-manager."""
+    # This is an adaptation of what subscription-manager's cli enforces:
+    # https://github.com/candlepin/subscription-manager/blob/main/src/rhsm/utils.py#L112
+
+    # It should look like it has a scheme followed by other information.
+    if not re.match("^[^:]+://.+", server_url):
+        # If there isn't a scheme, add one now
+        server_url = "https://%s" % server_url
+
+    url_parts = urllib.parse.urlsplit(server_url, allow_fragments=False)
+
+    if url_parts.scheme not in ("https", "http"):
+        raise ValueError(
+            "Subscription manager must be accessed over http or https.  %s is not valid" % url_parts.scheme
+        )
 
-        return output, ret_code
+    return url_parts
```
Method has no unit tests. If we want to make it private then we are missing a branch for the `ValueError`, can be covered with the exception handling where this method is called

abadger replied:
I've made this private and added unit_tests.  We're debating some more changes and cornercases in the serverurl so there might be a branch or two left to cover in the unit_tests but at least parse_url() (renamed) and the ValueError catch are being tested now.

SpyTec commented on convert2rhel/systeminfo.py:
```diff
@@ -359,6 +364,80 @@ def corresponds_to_rhel_eus_release(self):
+    return running
+
+
+def _get_dbus_status_from_systemd():
+    """Get DBus status from systemd."""
+    # Reloading, activating, etc will return None which means to retry
```
We're not retrieving the status, but rather checking whether it's running
```suggestion
def _check_dbus_service_running():
    """
    Check that DBus systemd service is running.
    :returns: true/false if status is determined. None if status is indeterminate
    """
```

abadger replied:
The three functions involved in checking whether dbus is running have been renamed here: https://github.com/oamg/convert2rhel/pull/540/commits/f0f7ffd0f5b51b824abc6e0f80f3fa5a46c8352c

SpyTec commented on convert2rhel/systeminfo.py:
```diff
@@ -359,6 +364,80 @@ def corresponds_to_rhel_eus_release(self):
+            if self.version.major <= 6:
+                status = _get_dbus_status_from_sysv()
+            else:
+                status = _get_dbus_status_from_systemd()
```
See review comment for `_get_dbus_status_from_systemd()`
```suggestion
                status = _check_dbus_service_running()
```

SpyTec commented on convert2rhel/unit_tests/__init__.py:
```diff
@@ -107,6 +109,21 @@ def safe_repr(obj, short=False):
     return result[:_MAX_LENGTH] + " [truncated]..."
 
 
+def get_pytest_mark(request, mark_name):
```
This name and docstring is confusing to me. I can't tell what it's doing until I read the code

abadger replied:
Would get_pytest_get_marker be better?  The pytest API uses `get_marker()` (old) and `get_closest_marker()` (new) for this functionality.
```suggestion
def get_pytest_marker(request, mark_name):
```

bocekm commented on convert2rhel/i18n.py:
```diff
@@ -0,0 +1,45 @@
+# Display locales
+#
+
+# These locales are about what we want to display to the user.  If we decide to localize, these
```
This is a matter of style: https://en.wikipedia.org/wiki/Sentence_spacing. For example:

> The text-editing environment in Emacs uses a double space following a period to identify the end of sentences unambiguously; the double-space convention prevents confusion with periods within sentences that signify abbreviations.

r0x0d replied:
Oh. I always thought it was something wrong with vim when doing `gq` for wrapping the text... Good to know it's a convnetion then.

abadger replied:
<nod>  Yeah.  I think if you're old enough, you were taught in (US schools at least) to use two spaces between sentences. (Love this headline: "[Nothing Says Over 40 Like Two Spaces after a Period!](https://www.cultofpedagogy.com/two-spaces-after-period" :-)  In addition to vim and emacs, google's android keyboard will insert a period if you type two concurrent spaces.  (Interestingly, I just googled for a way to set vim to change two spaces after period in python comments to single space and found out that PEP8 specifies two spaces after period due to the same tooling criteria, Emacs and vim use two spaces after period in their reflow tools.)

I'm committing all the places where Eric (or anyone has spotted) me using two spaces, but please be patient with me, I'm finding it very hard to retrain my fingers to only type one space after a period.

bocekm replied:
I'd say - feel free to use double-space. There's no harm in doing so :)

r0x0d commented on convert2rhel/subscription.py:
```diff
@@ -277,28 +287,131 @@ def args(self):
+        loggerinst.debug("Getting a subscription-manager RegisterServer object from dbus")
+        register_server = system_bus.get_object("com.redhat.RHSM1", "/com/redhat/RHSM1/RegisterServer")
+        loggerinst.debug("Starting a private DBus to talk to subscription-manager")
+        address = register_server.Start(
+            i18n.SUBSCRIPTION_MANAGER_LOCALE, dbus_interface="com.redhat.RHSM1.RegisterServer"
+        )
```
One question that strikes me is, this piece of code can't fail when we run it? I'm not sure about the dbus-library, but that could rise an unexpected exception?

abadger replied:
This code can fail.  You are probably asking why it isn't part of the `try: finally:` that starts on the following line?  The reason for that is we only need the `try: finally:` to ensure we tear down the private DBus server that we're creating in this line.  If this line fails, then there won't be a registration server to tear down.

I'm assuming that register_server.Start won't fail after it creates the private DBus server and before it returns to us.  If it does fail there, we could end up with the server still running.  If that's the case, we'd also have to make the `finally:` clause handle both the server having been started and the case where it had not yet been started.

The current subscription-manager starts the server, logs it, and then returns so it seems pretty safe to let subscription-manager's code worry about this cornercase.

r0x0d replied:
Oh, I see... It totally makes sense now, and you're correct, my question was regarding the `try: finally:` block. In my point of view, I would expect that the server start would fail and start a rollback immediately (or worse, raise an exception and prevent  us to do a rollback)

bocekm commented on convert2rhel/subscription.py:
```diff
@@ -120,24 +125,29 @@ def register_system():
-        attempt += 1
-    loggerinst.critical("Unable to register the system through subscription-manager.")
+        try:
+            registration_cmd()
```
```suggestion
            registration_cmd()
            loggerinst.info("System registration succeeded.")
```

bocekm reviewed (changes requested):
There's some weird behavior of the registration through dbus. I would expect it to raise exception if credentials are invalid. However it does not raise it and I get an error later on when trying to attach a subscription (at a stage where the system is supposed to be already registered).

On CentOS Linux 7 with `--username, `--password`, and `--pool` options used:

```
[08/10/2022 14:47:46] DEBUG - Getting a subscription-manager Register object from dbus
[08/10/2022 14:47:46] DEBUG - Registering via username/password: register --force --serverurl=subscription.rhsm.stage.redhat.com --username=oamg_acc
[08/10/2022 14:47:55] DEBUG - Shutting down private DBus instance
Attaching provided subscription pool ID to the system ...
[08/10/2022 14:47:55] DEBUG - Calling command 'subscription-manager attach --pool 8a99f9af7c95644d017cbb877eec04f8'
Invalid Credentials
Unable to sync syspurpose with server: HTTP error (401 - Unauthorized): Invalid Credentials
```

abadger commented on convert2rhel/unit_tests/__init__.py:
```diff
@@ -107,6 +109,21 @@ def safe_repr(obj, short=False):
+    Get a pytest mark from a request.
+
+    The pytest API to retrieve a mark changed between RHEL6 and RHEL7.  This function is
+    a compatibility shim to retrieve the value.
```
```suggestion
    a compatibility shim to retrieve the value.
    
    Use this function instead of pytest's `request.node.get_closest_marker(mark_name)` so that it will work on all versions of RHEL that we are targeting.
    .. seealso::
        * `pytest's get_closest_marker() function which this function wraps <https://docs.pytest.org/en/stable/reference/reference.html#pytest.nodes.Node.get_closest_marker>`
        * `A technique you might use where you would need to use this function to retrieve a mark's value. <https://docs.pytest.org/en/stable/how-to/fixtures.html#using-markers-to-pass-data-to-fixtures>`
```

abadger commented on packaging/convert2rhel.spec:
```diff
@@ -25,12 +25,17 @@ BuildRequires:  python%{python_pkgversion}-six
 BuildRequires:  python3-pexpect
 # rpm is being imported through utils.py
 BuildRequires:  python3-rpm
+# DBus is used by subscription.py
+BuildRequires:  python3-dbus
 %endif
 %if 0%{?rhel} && 0%{?rhel} <= 7
 BuildRequires:  pexpect
 BuildRequires:  rpm-python
+BuildRequires:  dbus-python
```
Interesting, okay.  At some point I'll have to look into why we need pexpect and rpm-python to build but not dbus.

bocekm commented on convert2rhel/subscription.py:
```diff
@@ -277,28 +297,144 @@ def args(self):
+                    self.org or "",
+                    self.username,
+                    self.password,
+                    EMPTY_DBUS_DICT,
```
Please replace the empty dict with the options dict enabling the `force` option.

abadger replied:
This has now been fixed in the PR.

bocekm commented:
Rebased to resolve a conflict coming from the merged https://github.com/oamg/convert2rhel/pull/552.

bocekm commented on convert2rhel/subscription.py:
```diff
@@ -245,9 +278,15 @@ def from_tool_opts(cls, tool_opts):
-            registration_attributes["server_url"] = tool_opts.serverurl
+        if tool_opts.rhsm_hostname:
+            loggerinst.debug("    ... using custom RHSM hostname")
+            self.rhsm_hostname = tool_opts.rhsm_hostname
+        if tool_opts.rhsm_port:
+            loggerinst.debug("    ... using custom RHSM port")
+            self.rhsm_port = tool_opts.rhsm_port
+        if tool_opts.rhsm_handler:
+            loggerinst.debug("    ... using custom RHSM handler")
+            self.rhsm_handler = tool_opts.rhsm_handler
```
```suggestion
            registration_attributes["rhsm_hostname"] = tool_opts.rhsm_hostname
        if tool_opts.rhsm_port:
            loggerinst.debug("    ... using custom RHSM port")
            registration_attributes["rhsm_port"] = tool_opts.rhsm_port
        if tool_opts.rhsm_handler:
            loggerinst.debug("    ... using custom RHSM handler")
            registration_attributes["rhsm_handler"] = tool_opts.rhsm_handler
```

abadger replied:
Thanks.  Fixed.

bocekm commented on convert2rhel/subscription.py:
```diff
@@ -120,28 +134,43 @@ def register_system():
 
 class RegistrationCommand(object):
-    def __init__(self, activation_key=None, org=None, username=None, password=None, server_url=None):
+    def __init__(
```
You removed `server_url`. Because of that you need to remove the line `self.server_url = server_url` from the `__init__()` body.

Also, some unit tests expect the parameter to be present, e.g. `test_calling_registration_command_with_serverurl`.

abadger replied:
Yeah, there were a lot of these left to be fixed.  I wanted to get the code up where   it could be looked at asap.  I think I've now fixed all of these places.

bocekm commented on convert2rhel/toolopts.py:
```diff
@@ -19,7 +19,9 @@
 import os
+import re
 import sys
+import urllib.parse
```
`No name 'parse' in module 'urllib'` on CentOS Linux 7.

r0x0d replied:
The CI is failing right now with the same errror

https://download.copr.fedorainfracloud.org/results/@oamg/convert2rhel/epel-7-x86_64/04726137-convert2rhel/builder-live.log.gz
https://download.copr.fedorainfracloud.org/results/@oamg/convert2rhel/epel-6-x86_64/04726137-convert2rhel/builder-live.log.gz

abadger replied:
Yeah, urllib.parse was introduced in python3.  I switched the imports to use `from six.moves import urllib` which should resolve this.

bocekm commented on convert2rhel/toolopts.py:
```diff
@@ -43,7 +45,9 @@ def __init__(self):
         self.enablerepo = []
         self.disablerepo = []
         self.pool = None
-        self.serverurl = None
```
There are some unit tests working with the toolopts.serverurl, like `test_custom_serverurl`.

abadger replied:
unit tests should now be updated to stop using serverurl

bocekm commented on convert2rhel/systeminfo.py:
```diff
@@ -359,6 +364,80 @@ def corresponds_to_rhel_eus_release(self):
+    """Get DBus status from SysVinit compatible systems."""
+    # None means the status should be retried because we weren't sure if it is turned off.
+    running = None
+    output, _ret_code = utils.run_subprocess(["/sbin/service", "messagebus", "status"])
```
`FileNotFoundError: [Errno 2] No such file or directory: '/sbin/service': '/sbin/service'` on CentOS Linux 8.

abadger replied:
I think I've tracked this down to the pretend_os fixture which populates system_info.  Inside of a container we might not have an init system so we need to hardcode a value for dbus_running rather than trying to use `service` or `systemctl` to probe for it.  Should be fixed now.

bocekm reviewed (changes requested):
Pylint/unit test errors.

r0x0d commented on convert2rhel/subscription.py:
```diff
@@ -245,9 +279,15 @@ def from_tool_opts(cls, tool_opts):
-            registration_attributes["server_url"] = tool_opts.serverurl
+        if tool_opts.rhsm_hostname:
+            loggerinst.debug("    ... using custom RHSM hostname")
+            self.rhsm_hostname = tool_opts.rhsm_hostname
+        if tool_opts.rhsm_port:
+            loggerinst.debug("    ... using custom RHSM port")
+            self.rhsm_port = tool_opts.rhsm_port
+        if tool_opts.rhsm_prefix:
+            loggerinst.debug("    ... using custom RHSM prefix")
+            self.rhsm_prefix = tool_opts.rhsm_prefix
```
`self` is not defined in this function

r0x0d commented on convert2rhel/subscription.py:
```diff
@@ -245,9 +279,15 @@ def from_tool_opts(cls, tool_opts):
-            registration_attributes["server_url"] = tool_opts.serverurl
+        if tool_opts.rhsm_hostname:
+            loggerinst.debug("    ... using custom RHSM hostname")
+            self.rhsm_hostname = tool_opts.rhsm_hostname
+        if tool_opts.rhsm_port:
+            loggerinst.debug("    ... using custom RHSM port")
+            self.rhsm_port = tool_opts.rhsm_port
+        if tool_opts.rhsm_prefix:
+            loggerinst.debug("    ... using custom RHSM prefix")
+            self.rhsm_prefix = tool_opts.rhsm_prefix
```
`self` is not defined here, it's a `classmethod`

```suggestion
            cls.rhsm_hostname = tool_opts.rhsm_hostname
        if tool_opts.rhsm_port:
            loggerinst.debug("    ... using custom RHSM port")
            cls.rhsm_port = tool_opts.rhsm_port
        if tool_opts.rhsm_prefix:
            loggerinst.debug("    ... using custom RHSM prefix")
            cls.rhsm_prefix = tool_opts.rhsm_prefix
```

abadger replied:
Michal actually had a slightly different solution here: https://github.com/oamg/convert2rhel/pull/540#discussion_r943563943

It's a little tricky, if you set `cls.rhsm_hostname`, then you are setting the attribute on the class.  And then when the instance of the class is initialized in `__init__()`, the instance attribute will mask the class attribute.  (ie: this would end up being None because it is set to None by `__init__()`).  Adding the values to `registration_attributes` which are the parameters which are passed into `__init__()` is the right way to fix my mistake.

r0x0d commented on convert2rhel/toolopts.py:
```diff
@@ -464,5 +496,43 @@ def options_from_config_files(cfg_path=None):
+    if not re.match("https?://[^/]+", serverurl):
+        # Anthing that looks like a malformed scheme is immediately discarded
+        if re.match("^[^:]+:/.+", serverurl):
+            raise ValueError("serverurl looks unparsable. Make sure it starts with http://HOST or https://HOST")
```
Does anyone know a way to rephrase this? The sentence starting with a lowercase seems strange. Maybe:

```suggestion
            raise ValueError("Problem while parsing serverurl. Make sure it starts with http://HOST or https://HOST")
```

abadger replied:
```suggestion
            raise ValueError("Unable to parse --serverurl.  Make sure it starts with http://HOST or https://HOST")
```

r0x0d commented on convert2rhel/unit_tests/subscription_test.py:
```diff
@@ -761,37 +766,118 @@ def test_args(self, registration_kwargs):
-            expect_script=(("[Pp]assword: ", "a password\n"),),
-            print_cmd=False,
+        assert run_subprocess_mocked.call_count == 1
+        print(run_subprocess_mocked.call_args.args[0])
```
I think you forgot a print here

abadger replied:
Thanks.  That code was failing on centos8 so I was trying to see what was going on.  I think I have it fixed and all the prints have been removed.

bocekm commented on convert2rhel/unit_tests/subscription_test.py:
```diff
@@ -494,6 +494,10 @@ def test_register_system_fail_interactive(self, tool_opts, monkeypatch, caplog,
     def test_register_system_keyboard_interrupt(self, tool_opts, monkeypatch, caplog, mocked_rhsm_register):
         """Test that we stop retrying if the user hits Control-C.."""
+
+        print(os.environ)
```
Unfortunately there's nothing more than in a docker image when executed locally:
```
 {'LANG': 'C', 'APP_MAIN_DEPS': 'python-six     dbus-python     pexpect', 'LC_ALL': 'C', 'PYTHON': 'python2', 'HOSTNAME': 'c80479088321', 'PYTHONDONTWRITEBYTECODE': '1', 'PYTEST_CURRENT_TEST': 'convert2rhel/unit_tests/subscription_test.py::TestRegisterSystem::test_register_system_keyboard_interrupt (call)', 'APP_DEV_DEPS': 'requirements/centos7.requirements.txt', 'URL_GET_PIP': 'https://bootstrap.pypa.io/pip/2.7/get-pip.py', 'HOME': '/home/app', 'PIP': 'pip', 'PATH': '/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin'}
 ```

abadger replied:
mbocekm later spotted that this test was calling `register_system()` twice and one of those was outside of the `with pytest.raises()`.  I've removed the extra `register_system()` and the unit_tests now pass.

bocekm commented:
The integration tests fail on `test_terminate_on_subscription_prompt` (https://github.com/oamg/convert2rhel/blob/main/tests/integration/tier0/test-check-rollback-handling/test_check_rollback_handling.py#L150):
```
System registration failed with error: com.redhat.RHSM1.Error: {"exception": "DBusException", "message": "This system is already registered", "severity": "error"}
```

r0x0d commented on convert2rhel/subscription.py:
```diff
@@ -120,33 +134,57 @@ def register_system():
+        # The subscription-manager DBus API has a force parameter but there's
+        # a bug in susbcription-manager where that doesn't take effect.
+        # Explicitly unregistration here to workaround that.
+        loggerinst.info("Try to unrregister the system to clear the server's state for our registration.")
```
Typo
```suggestion
        loggerinst.info("Try to unregister the system to clear the server's state for our registration.")
```

r0x0d commented on convert2rhel/subscription.py:
```diff
@@ -277,28 +338,125 @@ def args(self):
+            register_object = private_bus.get_object("com.redhat.RHSM1", "/com/redhat/RHSM1/Register")
+
+            if self.password:
+                loggerinst.debug("Registering via username/password: %s" % " ".join(hide_secrets(self.args)))
```
A minor thing here: 

I feel that this should be a `info` instead of a `debug` to actually inform the user that we are registering the system. What you think?

bocekm replied:
I don't think it's necessary. This is what they see now.

``` subscription-manager registration info ... 
09:45:31                 out:     ... activation key not found, using given username and password
09:45:31                 out:     ... username detected
09:45:31                 out:     ... password detected
```

If the credentials are incorrect, they'll get the `"Invalid username or password."` error. Then they can look at the convert2rhel.log file to see what username they used or even better look at how they called convert2rhel so they can see even the password they used.

r0x0d replied:
Makes sense to me! Thanks

abadger commented on convert2rhel/subscription.py:
```diff
@@ -120,33 +137,86 @@ def register_system():
+            # subscription-manager Registration API reports that the system is
+            # already registered. For the short term, we are going to stop the
+            # rhsm service to work around this issue:
+            loggerinst.info("Stopping rhsm so that registration does not think that the host is already registered.")
```
I've changed this in the latest push.

bocekm commented on convert2rhel/toolopts.py:
```diff
@@ -345,7 +348,36 @@ def _process_cli_options(self):
+                if url_parts.port:
+                    # urllib.parse.urlsplit() convert2 this into an int but we
+                    # always use it as a str
+                    tool_opts.rhsm_port = str(url_parts.port) if url_parts.port else ""
```
```suggestion
                    tool_opts.rhsm_port = str(url_parts.port)
```
This line already is under the condition `if url_parts.port:`.

abadger replied:
Good point.  I have a vague recollection that the `else ""` might be needed so I changed the code to not do `if url_parts.port` first.

abadger commented on convert2rhel/subscription.py:
```diff
@@ -277,28 +377,125 @@ def args(self):
+        """
+        # Note: dbus doesn't understand empty python dicts. Use dbus.Dictionary({}, signature="ss")
+        # if we need one in the future.
+        REGISTER_OPTS_DICT = dbus.Dictionary({"force": True}, signature="ss")
```
`ofrce` in `REGISTER_OPTS` not working seems to be a bug.  So it seems like it would be better to leave it in?

abadger replied:
Comment in the code has been updated to link to the bugzilla.

bocekm replied:
Ok, let's leave this in.

bocekm commented on convert2rhel/subscription.py:
```diff
@@ -277,28 +376,125 @@ def args(self):
+        """
+        # Note: dbus doesn't understand empty python dicts. Use dbus.Dictionary({}, signature="ss")
+        # if we need one in the future.
+        REGISTER_OPTS_DICT = dbus.Dictionary({"force": True}, signature="ss")
```
The signature is incorrect. The registration fails on OL 8.6 with:
```
ERROR:dbus.connection:Unable to set arguments ('', 'username', 'password', dbus.Dictionary({'force': True}, signature=dbus.Signature('ss')), dbus.Dictionary({'host': 'subscription.rhsm.redhat.com'}, signature=dbus.Signature('ss')), 'C') according to signature None: <class 'TypeError'>: Expected a string or unicode object
```

abadger replied:
This should be fixed.  Correct signature is `dbus.Dictionary({}, signature="sv", variant_level=1)`.  If that doesn't work (It's possible I don't understand variant_level) then we can removethe register_opts dict for this PR and revisit adding it back in later.

bocekm commented on convert2rhel/subscription.py:
```diff
@@ -22,7 +22,12 @@
 
-from convert2rhel import backup, pkghandler, utils
+import dbus
+import dbus.connection
```
```suggestion
```
Unless it's about some kind of optimization, it isn't necessary.

abadger replied:
I like to have this because whether this works: `import dbus ; dbus.connection` is an implementation detail of the dbus module.  It will work if `dbus/__init__.py` has something like:
``` python
from dbus import connection
```
If `dbus/__init__.py` does not have something like that line, then `import dbus.connection` would be necessary in our code.

bocekm commented on convert2rhel/subscription.py:
```diff
@@ -96,14 +114,11 @@ def unregister_system():
-            ret_code,
-            output,
-        )
+        raise UnregisterError("System unregistration failed with return code %d and message:\n%s" % (ret_code, output))
```
```suggestion
        raise UnregisterError("System unregistration result:\n%s" % output)
```
Because most customers will see this warning (before registering), I'd ease out the wording.

bocekm commented on convert2rhel/subscription.py:
```diff
@@ -277,28 +377,173 @@ def args(self):
+            loggerinst.debug("Shutting down private DBus instance")
+            register_server.Stop(i18n.SUBSCRIPTION_MANAGER_LOCALE, dbus_interface="com.redhat.RHSM1.RegisterServer")
+
+        loggerinst.info("Successfully registered")
```
I think you added this later after you added `loggerinst.info("System registration succeeded.")`.

Now two similar messages are printed:
```
Successfully registered
System registration succeeded.
```

abadger replied:
I agree.  I think I added this as I was writing unit_tests and realized that there isn't much to go on for a successful exit.  I'll remove it and depending on returning None as meaning success here.

bocekm commented on convert2rhel/subscription.py:
```diff
@@ -120,33 +136,94 @@ def register_system():
+    cmd = ["/bin/systemctl", "stop", "rhsm"]
+    if system_info.version.major <= 6:
+        # RHEL6 used upstart.  After that, we use systemd
+        cmd = ["service", "rhsm", "stop"]
```
```
[08/17/2022 20:01:59] DEBUG - Calling command 'service rhsm stop'
Stopping RHSM failed with code: 1; output: rhsm: unrecognized service
```

I looked at the services (`service --status-all`) and there's just `rhsmcertd` which I'm not sure makes sense stopping (service description: `Enable periodic update of entitlement certificates.`).
```
rpc.svcgssd is stopped
rpc.mountd is stopped
nfsd is stopped
rpc.statd (pid  1162) is running...
master (pid  1326) is running...
rdisc is stopped
restorecond is stopped
rhsmcertd is stopped
rpcbind (pid  1140) is running...
rpc.gssd is stopped
rpc.idmapd is stopped
rpc.svcgssd is stopped
rsyslogd (pid  1111) is running...
sandbox is stopped
saslauthd is stopped
openssh-daemon (pid  19422) is running...
```

abadger replied:
Hmmm.... Okay, so I just ran our dbus test script on CentOS6 and then looked at the process list and found this:
> 3485 ?        Sl     0:00 python /usr/libexec/rhsm-service

I think that rhsm-service is started on-demand when something uses the DBus API.  On systemd hosts (CentOS7 and up), this is handled by systemd which waits for a socket to be opened and then uses that to launch rhsm-service so that it can handle the request.  On centos6, the subscription-manager code must do this itself and they haven't made a SysVinit script to start the service there.  If we need to stop the service, I think we'll need to find the rhsm-service process and send it a `kill -TERM`.

I haven't been able to provoke the system is already registered error with the dbus reg.py script so I'm not certain whether RHEL6 has this problem or not.

Here's what I propose... I will take out the code path that operates on CentOS-6.  If you want to look into whether it's needed on CentOS 6 (or if we want to wait to see if a user runs into a problem) we can implement sending a signal to rhsm-service then.  Does that sound good?

bocekm replied:
I was also thinking that on RHEL6-like systems we may not need to stop the service because it works differently there. So leaving the service-stopping out for OS 6 sounds good to me. Thanks for looking into it.

bocekm reviewed (approved):
Tested manually and works well. Code looks good. My previous review doesn't talk about anything critical.

bocekm commented on convert2rhel/subscription.py:
```diff
@@ -201,8 +201,12 @@ def _stop_rhsm():
+        # there isn't an init script to stop it.  If we find we need to stop it, because we're
+        # etting "machine is already registered" errors there, then we'll need to look for
+        # rhsm-service in the process list and send it the TERM signal.
+        loggerinst.info("Skipping RHSM service shutdown on CentOS 6.")
```
```suggestion
loggerinst.info("Skipping RHSM service shutdown on {0} {1}.".format(system_info.name, system_info.version.major))
```
The message would be confusing on Oracle Linux 6. But it's nothing major.

abadger commented:
With the latest changes, OEL-8.6 has passed integration tests.  I'm going to update the log message that Michal mentioned and then make sure all of the integration tests are restarted successfully.

danmyway commented:
The `check_cve` test got stuck for some reason.
Yesterday I have sent the test tiers separately to the testing farm, where it passed. 
Results to be seen here: http://artifacts.osci.redhat.com/testing-farm/2f67bf47-dee8-4b73-9671-22fa75563ba5/
Specifically the `check_cve` test here http://artifacts.osci.redhat.com/testing-farm/2f67bf47-dee8-4b73-9671-22fa75563ba5/work-check_cvejwWpdh/log.txt
Also I have run the test locally and it passed

danmyway commented:
Another successful separate run http://artifacts.osci.redhat.com/testing-farm/1ce38c8c-cd01-4c2c-8c85-c5c71f1b83cf/

bocekm commented:
After talking with @danmyway about details of the integration test runs on OL 8.6, all of them passed when he scheduled them manually outside the GitHub CI. He considers the PR verified.
